src/design_docs/async_stack_traces.md

```diff
@@ -0,0 +1,705 @@
+
+The second is that rather than seeing a call to `common::foo` or similar, we see `common::foo::generator$0`.
+
+These two are relatively minor issues.
+Fixing them may not be desirable.
+
+## Proposed Solutions
+
+There are several improvements we could make that would improve the state of backtraces.
+
+### Allow runtimes to trim startup code
+
+This solution probably has the highest impact to effort ratio.
+The core of the idea is to move the call to `__rust_begin_short_backtrace` when running under an async executor.
+Most likely the way this would happen is to have an alternate startup path that programs can opt in to (how to actually make this work is left as an exercise for the reader).
+The alternate path would not call `__rust_begin_short_backtrace` in [`rt.rs`], but would instead expect the program to make sure to call it at the appropriate time.
+Then, async runtimes that provide a macro such as `#[tokio::main]` or `#[async_std::main]` would arrange to start up through this alternate path and call `__rust_begin_short_backtrace` shortly before invoking user code.
+
+[`rt.rs`]: https://github.com/rust-lang/rust/blob/master/library/std/src/rt.rs
+
+Note that this approach does not help in cases where users create and launch the runtime manually rather than using a library-provided macro.
+It does help in the most common cases, however, and when users take a more manual approach they would also be able to control where the short stack trace starts.
+
+### Allow trimming of internal frames
+
+There are several ways to do this, with varying levels of implementation effort and cost.
+
+The conceptually simplest is to allow multiple `__rust_begin_short_backtrace`/`__rust_end_short_backtrace` pairs.
+Implementing this could be done almost entirely with changes to [`_print_fmt`].
+This approach has some serious drawbacks.
+First, it requires a lot of work from library authors to annotate each transition point between user and library code.
+Second, `__rust_begin_short_backtrace`/`__rust_end_short_backtrace` are built to defeat inlining to ensure they show up as a frame on the stack.
+This is fine when the functions are only called during process startup and once when a panic starts, but it would likely be prohibitively expensive if interspersed between every async function call.
+
+[`_print_fmt`]: https://github.com/rust-lang/rust/blob/master/library/std/src/sys_common/backtrace.rs#L52
+
+It might be possible instead to include more information in the debugging symbols.
+For example, we might be able to add a flag indicating that a certain function should be hidden from the backtrace.
+To do this, we would first need to make sure existing debugging formats such as DWARF and PDB are able to encode such information.
+If there is already support for this, then it's likely debuggers would benefit as well since they would also be able to display trimmed backtraces.

```

Comment:
This wouldn't work for builds without debug symbols, presumably.

Reply:
Good point. I'll add a note making that explicit.